src/rgw/driver/rados/topic_migration.cc

```diff
@@ -100,12 +100,27 @@ int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
       return r;
     }
 
-    if (bucket->get_marker() != marker || bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+    if (bucket->get_marker() != marker) {
       break;
     }
 
+    rgw_pubsub_bucket_topics v2_bucket_topics;
+    if(bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
```

Comment:
we're in `migrate_notification()` to process a rados object of the form "pubsub.{tenant}.bucket.{name}/{marker}"

why would we expect an existing RGW_ATTR_BUCKET_NOTIFICATION attr? what other `migrate_notification()` call could possibly generate an attr on that bucket that we'd need to merge here?

Reply:
i don't think it is a question on wheter we expect that to be a useful case or not.
it *can* happen (see: test_ps_s3_data_path_v2_mixed_migration), and we need a consistent way of handling it.
doing the merge (where we prefer the v2 notifications) is consistent with what we do for topics

Reply:
isn't skipping buckets that already have an RGW_ATTR_BUCKET_NOTIFICATION attr also a consistent way to handle that case?

disabling v2 and writing conflicting v1 metadata isn't a case that we want to support. do we really need to write tests that rely on that? i really don't like that we're making the migration code more complicated just because it's hard to test otherwise

Reply:
in order to be consistent with topics, we wouls need to check that we have the new topics objects, and not migrate the v1 topics if we see any.
IMO, this adds more complexity than the merge

Reply:
> in order to be consistent with topics, we wouls need to check that we have the new topics objects, and not migrate the v1 topics if we see any.

can you please expand on the consistency issue you're describing? are we still talking about internal topics here?

Reply:
this is a different issue. regarding the internal (auto-generated) topics, we won't migrate them. had to change one thing in `rgw_notify.cc` to make sure that we are able to send async notifications even without the auto-generated topics.

regarding consistency. in the `migrate_topics()` function we handle the case that we are trying to migrate the v1 topics object, in an environment that has v2 topics (regardless of why we got there, it is possible). in that cae, we do not override the topics that we already migrated, but keep migrating other v1 topics.
to be consistent with this behavior in the case of notifications, we should keep migrating v1 notifications even if there are v2 notifications that were already migrated (but make sure we don't override).

Reply:
thanks @yuvalif, that makes sense

one small request though - if `migrate_notification()` finds an existing attr but `merge()` doesn't add anything new, please avoid writing the bucket info back. that way we don't have to invalidate the metadata cache or bump the cls version